_posts/2020-07-08-event-driven-functions-with-openfaas-and-nats.md

```diff
@@ -0,0 +1,152 @@
+
+## NATS
+
+[NATS](https://nats.io) is a simple, secure and high performance open source messaging system for cloud native applications, IoT messaging, and microservices architectures. In this blog post, I will be using NATS publish-subscribe concept where publishers publishes a message on a topic/subject and subscribers consumes that message by subscribing to that subject (sometimes called topics in other event systems).
+
+![NATS Publish Subscribe](/images/2020-openfaas-nats/nats-publish-subscribe.png "https://docs.nats.io/nats-concepts/pubsub")
+
+## Pre-requisite Installation
+
+* [k3d](https://github.com/rancher/k3d) to create a [k3s](https://github.com/rancher/k3s) cluster. k3s is lightweight distribution of kubernetes from Rancher Labs.
+* [arkade](https://github.com/alexellis/arkade) to install OpenFaaS, nats-connector, kubectl and faas-cl. `arkade` is simple CLI tool to install helm charts and apps to your cluster in one command.
+
+> Please make sure you have these tools installed before you proceed next.
+
+## Install Components
+
+#### Create a kubernetes cluster
+
+`k3d` is a tool to run `k3s` in docker. Create a single node `k3s` cluster. You can also use other tools like [KinD](https://github.com/kubernetes-sigs/kind)
+
+```
+k3d create
+```
+
+Follow instructions returned in the output of this command to configure KUBECONFIG to access the cluster using kubectl.
+
+#### Install OpenFaaS 
+Install OpenFaaS using `arkade`
+
+```
+arkade install openfaas
+```
+
+#### Install NATS connector using arkade
+`nats-connector` is connector which invokes OpenFaaS function when a message is sent to a NATS topic.
+
+```
+arkade install nats-connector 
+```
+
+> NATS comes with OpenFaaS installation, so we are not setting up NATS here
+
+#### Install kubectl and faas-cli
+
+install `kubectl` and `faas-cli` using `arkade` if they are not already installed.   
+
+```
+arkade get kubectl
+```
+
+```
+arkade get faas-cli
+```
+
+#### Login to OpenFaaS gateway using CLI
+
+Port forward the gateway service to access it. 
+
+```
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Get the password
+
+```
+PASSWORD=$(kubectl get secret -n openfaas basic-auth -o jsonpath="{.data.basic-auth-password}" | base64 --decode; echo)
+```
+
+Login to the gateway
+
+```
+echo -n $PASSWORD | faas-cli login --username admin --password-stdin
+```
+
+
+## Deploy functions
+`nats-connector` repo comes with test functions to verify installation. It has two OpenFaas functions.
+
+`receive-message` function subscribes to NATS topic nats-test. On every new message on nats-test topic, receive-message will be invoked.
+
+`publish-message` function publishes message to nats-test topic.
+
+Deploy `receive-message` function
+
+```
+faas-cli deploy --name receive-message --image openfaas/receive-message:latest \
```

Comment:
Could we have a stack.yml for this instead please?

Reply:
Or if this is in a git repo (I think it is), then let's clone and deploy with `faas-cli deploy`

Reply:
updated